TIM/TIM Core Code/man/man3/subsys.3
.TH "subsys" 3 "Tue Jan 26 2016" "Version 0.1" "P21451_TIM" \" -*- nroff -*-
.ad l
.nh
.SH NAME
subsys \- Subsystem Module
.SS "Data Structures"

.in +1c
.ti -1c
.RI "struct \fBversion_mmb_t\fP"
.br
.ti -1c
.RI "union \fBversion_t\fP"
.br
.RI "\fIversion union \fP"
.in -1c
.SS "Macros"

.in +1c
.ti -1c
.RI "#define \fBSYSTEM_VERSION\fP   0x03010001u"
.br
.RI "\fIsubsystem module version number \fP"
.ti -1c
.RI "#define \fBRECEIVE_MAX_LENGTH\fP   64"
.br
.RI "\fIdefault length of receive line for commands \fP"
.ti -1c
.RI "#define \fBRECEIVE_START_CHAR\fP   '$'"
.br
.RI "\fIstart character for command lines \fP"
.ti -1c
.RI "#define \fBRECEIVE_STOP_CHAR\fP   '\\r'"
.br
.RI "\fIstop character for command lines \fP"
.ti -1c
.RI "#define \fBRECEIVE_MAX_ARGC\fP   8"
.br
.RI "\fImax number of arguments after the command name \fP"
.ti -1c
.RI "#define \fBGetLogTimestamp\fP()   \fBTimeNow\fP()"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "void \fBLogStr\fP (char *str,\&.\&.\&.)"
.br
.ti -1c
.RI "void \fBLogMsg\fP (uint8_t subsystem_id, char *str,\&.\&.\&.)"
.br
.ti -1c
.RI "uint8_t \fBSubsystem_Init\fP (char *name, \fBversion_t\fP version, void(*callback)(int argc, char *argv[]))"
.br
.ti -1c
.RI "void \fBSubsystem_RegisterCallback\fP (uint8_t subsystem_id, void(*callback)(int argc, char *argv[]))"
.br
.ti -1c
.RI "char * \fBGetSubsystemName\fP (uint8_t subsystem_id)"
.br
.ti -1c
.RI "void \fBLog_EchoOn\fP (void)"
.br
.ti -1c
.RI "void \fBLog_EchoOff\fP (void)"
.br
.ti -1c
.RI "uint8_t \fBLog_GetEcho\fP (void)"
.br
.ti -1c
.RI "void \fBLog_Header\fP (uint8_t subsystem_id)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Created on: Mar 12, 2014 
.PP
\fBAuthor:\fP
.RS 4
: Michael
.RE
.PP
\fBVersion:\fP
.RS 4
2014\&.03\&.26 changed SystemTick to SubsystemTick so the Task Management Module can own \fBSystemTick()\fP 
.PP
2\&.0\&.1 added version, setup SYSTEM subsystem, changed warning messages to priority WARNING and set default priority to ERROR to suppress messages 
.PP
3\&.1\&.1 removed log levels and redueced to mute or unmute\&. Moved receiver functionality to UART module\&. Changed callback to take argc argv inputs and added parsing to create argc and argv\&. 
.RE
.PP
SubSystem module command interface
If bi-directional communication is available (e\&.g\&. UART) then the user can interface with the SubSystem module in real time using the module's command interface\&.
.PP
A command can be sent to the module in the following format: 
.PP
.nf
$command -f0

.fi
.PP
 Where the $ indicates the start of the command, command is the name of the command and -f8 is a flag with a value of 8\&. 
.br
Valid system flags are \fC-s\fP, \fC-l\fP and \fC-g\fP 
.PP
The commands available are:
.PP
\fCversion\fP (or \fCver\fP) - will output a list of subsystems and their version\&. No flags are applicable\&.
.PP
\fClevel\fP (or \fClev\fP) - if no flags are given \fClevel\fP will output level a list of valid level codes and their names as used in the log messages\&. If a \fC-l\fP flag is given but not a \fC-s\fP flag then the level specified with the \fC-l\fP flag is set as the global level\&. If the \fC-l\fP and \fC-s\fP flags are given then set the level specified with the \fC-l\fP flag for the subsystem specified with the \fC-s\fP flag\&. If a \fC-g\fP flag is given it will set the global log level to the specified value\&. The global log level will for any message from any subsystem to be logged if its level is at the global level or below\&.
.PP
\fCsubsystem\fP (or \fCsub\fP or \fCsys\fP or \fCsystem\fP) - will output a list of valid subsystem codes and their associated subsystem names and their current log level setting as used by the log message functions\&. The \fC-s\fP flag can be used to return only information about the specified subsystem\&.
.PP
Example usage: 
.PP
.nf
$version

$sub

$level

$level -l0

$level -s0 -l2

.fi
.PP
.PP
In the above example:
.IP "\(bu" 2
output the subsystems and their versions
.IP "\(bu" 2
output the subsystems and their codes
.IP "\(bu" 2
output the log levels and their codes
.IP "\(bu" 2
set the global log level to 0 (OFF)
.IP "\(bu" 2
set the log level for subsystem 0 (SYSTEM) is set to 2 (ERROR)\&.
.PP
.PP
The end result is that no messages would be logged unless they were SYSTEM messages with a priority level greater than or equal to ERROR (note higher priority is lower numerically)\&.
.PP
Additionally commands may be forwarded to the callback of compatible subsystems using the following format 
.PP
.nf
$12 command <anything goes>
$MUH play

.fi
.PP
 In the first line subsystem index 12 is sent the command string starting after the space\&. In the second line a module named MUH is sent a command 'play'
.PP
\fBTodo\fP
.RS 4
MM Update this comment block 
.RE
.PP

.SS "Subsystem module initialization"
Below are three example of how a subsystem could initialize itself\&.
.PP
.PP
.nf
// define the version
#define TASK_VERSION (version_t)0x01010014u

// initialize module to log EVERYTHING and to use "task" to refer
// to this subsystem in output messages
SubsystemInit(TASK, EVERYTHING, "task", TASK_VERSION);

// or another way to do the same thing

#define TASK_VERSION_MAJOR 1
#define TASK_VERSION_MINOR 1
#define TASK_VERSION_BUILD 20

version_t task_version;
task_version\&.major = TASK_VERSION_MAJOR;
task_version\&.minor = TASK_VERSION_MINOR;
task_version\&.build = TASK_VERSION_BUILD;

SubsystemInit(TASK, EVERYTHING, "task", task_version);

// or to do it all in one line
uint8_t task_id;
task_id = SubsystemInit(EVERYTHING, "task", (version_t)0x01010014u);
LogMsG(task_id, WARNING, "Crap hit the fan");
.fi
.PP
.PP
\fBTodo\fP
.RS 4
MM Update this comment block 
.RE
.PP

.SH "Macro Definition Documentation"
.PP 
.SS "#define GetLogTimestamp()   \fBTimeNow\fP()"
GetLogTimestamp must be defined so that it returns a integer (up to 32 bits) timestamp 
.PP
Definition at line 156 of file subsys\&.h\&.
.SS "#define RECEIVE_MAX_ARGC   8"

.PP
max number of arguments after the command name 
.PP
Definition at line 27 of file subsys\&.h\&.
.SS "#define RECEIVE_MAX_LENGTH   64"

.PP
default length of receive line for commands 
.PP
Definition at line 23 of file subsys\&.h\&.
.SS "#define RECEIVE_START_CHAR   '$'"

.PP
start character for command lines 
.PP
Definition at line 25 of file subsys\&.h\&.
.SS "#define RECEIVE_STOP_CHAR   '\\r'"

.PP
stop character for command lines 
.PP
Definition at line 26 of file subsys\&.h\&.
.SS "#define SYSTEM_VERSION   0x03010001u"

.PP
subsystem module version number 
.PP
Definition at line 20 of file subsys\&.h\&.
.SH "Function Documentation"
.PP 
.SS "char* GetSubsystemName (uint8_t subsystem_id)"
Return a pointer to a string corresponding to the name of the subsystem
.PP
The name returned is the one set by SubsystemInit()
.PP
\fBParameters:\fP
.RS 4
\fIsubsystem_id\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
- pointer to a null terminated string corresponding to the name of the subsystem 
.RE
.PP

.PP
Definition at line 85 of file subsys\&.c\&.
.SS "void Log_EchoOff (void)"
Turn echo feature off 
.PP
Definition at line 349 of file subsys\&.c\&.
.SS "void Log_EchoOn (void)"
Turn echo featuren on (default is off) 
.PP
Definition at line 345 of file subsys\&.c\&.
.SS "uint8_t Log_GetEcho (void)"
Get status of echo setting
.PP
\fBReturns:\fP
.RS 4
echo setting, 1 if echo is on 
.RE
.PP

.PP
Definition at line 353 of file subsys\&.c\&.
.SS "void Log_Header (uint8_t subsystem_id)"
Log header (timestamp and subsystem name)
.PP
\fBParameters:\fP
.RS 4
\fIsubsystem_id\fP subsystem id (index) 
.RE
.PP

.PP
Definition at line 294 of file subsys\&.c\&.
.SS "void LogMsg (uint8_t subsystem_id, char * str,  \&.\&.\&.)"
Logs the message at the pointer (str) with a timestamp and subsystem name
.PP
Before logging the message the function will check the current log setting of the subsystem and to determine if the message should be logged
.PP
This function is implemented using Push_vprintf\&. See \fBPush_printf()\fP for supported flags/features\&.
.PP
Will log the string to the buffer defined by SUSSYS_UART
.PP
\fBParameters:\fP
.RS 4
\fIsubsystem_id\fP subsystem id 
.br
\fIstr\fP pointer to message to log 
.br
\fI\&.\&.\&.\fP variable number of replacement parameters for the str string
.RE
.PP
Example usage: 
.PP
.nf
1 LogMsg(sys\&.id, "oops I crapped my pants");
2 LogMsg(sys\&.id, "System Index %d, System Name %s\&.", sys\&.id, GetSubsystemName(SYSTEM));

.fi
.PP
 
.PP
Definition at line 275 of file subsys\&.c\&.
.SS "void LogStr (char * str,  \&.\&.\&.)"
Logs the null terminated string at the pointer (str)
.PP
Same as \fBLogMsg()\fP without the header in the beginning and without the CRLF at the end\&.
.PP
This function is implemented using Push_vprintf\&. See \fBPush_printf()\fP for supported flags/features\&.
.PP
Will log the string to the buffer defined by LOG_BUF (typically tx0)
.PP
\fBParameters:\fP
.RS 4
\fIstr\fP pointer to string to log 
.br
\fI\&.\&.\&.\fP variable number of replacement parameters for the str string
.RE
.PP
Example usage: 
.PP
.nf
1 LogStr("oops I crapped my pants");
2 LogStr("System Index %d, System Name %s\&.", SYSTEM, GetSubsystemName(SYSTEM));

.fi
.PP
 
.PP
Definition at line 264 of file subsys\&.c\&.
.SS "uint8_t Subsystem_Init (char * name, \fBversion_t\fP version, void(*)(int argc, char *argv[]) callback)"
Initialize settings for a subsystem - critical for proper logging and command interface
.PP
If a module/subsystem uses logging it should call this function with the appropriate inputs when the subsystem is initializing\&.
.PP
\fBReturns:\fP
.RS 4
subsystem index 
.RE
.PP
\fBParameters:\fP
.RS 4
\fIname\fP pointer to name of the subsystem (recommended to make the name 8 characters or less) 
.br
\fIversion\fP software version of subsystem, see \fBversion_t\fP for more info 
.br
\fIcallback\fP callback function to be called when the user inputs a command in the form of '$name var1 var2 var3\&.\&.\&.'\&. Where name is the name passed to this function\&. The callback will be passed the number of arguments, \fCargc\fP , and a array of pointers to the argument strings, \fCargv\fP\&. 
.RE
.PP

.PP
Definition at line 102 of file subsys\&.c\&.
.SS "void Subsystem_RegisterCallback (uint8_t subsystem_id, void(*)(int argc, char *argv[]) callback)"
Register a callback function for a subsystem
.PP
When a command is received by the logging module for the subsystem \fCsys\fP the \fCcallback\fP function will be executed and passed the number of arguments \fCargc\fP and a array (vector) of pointer to the argument strungs \fCargv\fP\&.
.PP
The callback is set by \fBSubsystem_Init()\fP, this function can be used to update the callback\&.
.PP
\fBParameters:\fP
.RS 4
\fIsubsystem_id\fP - subsystem to register the callback for 
.br
\fIcallback\fP - function pointer to the function to run when a command is received for the subsystem\&. 
.RE
.PP

.PP
Definition at line 113 of file subsys\&.c\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for P21451_TIM from the source code\&.


TIM/TIM Core Code/subsys.c
/*
 * subsys.c
 *
 *  Created on: Mar 12, 2014
 *      Author: Michael M
 */
#include "system.h"
#include "macros.h"
#include <string.h>
#include <strings.h>

#ifndef SUBSYS_UART
#error "You must define what UART number the terminal is using in system.h: #define SUBSYS_UART num"
#endif

#ifndef SUBSYS_H_
#error "Please include subsys.h in system.h"
#endif

//if(sys.echo) UART_WriteByte(0, c);
#ifndef MAX_SUBSYSTEMS
#define MAX_SUBSYSTEMS 10
#endif

#define NUM_ADDRESS

void ReceiveChar(char c);
void ProcessMsg(void);
void LogLevels(void);
void LogVersions(void);
void LogSubsystem(int16_t s);
static void SystemCallback(int argc, char *argv[]);

/** Subsystem struct to organize name, level, and version for each subsystem.
 *
 * Note: this struct is kept local to this file since the user should use the
 * following methods to set and update the subsystem information:
 * @see SubsystemLogPriority()
 * @see SubsystemInit()
 *
 */
typedef struct {
    char * name; /**< pointer to name of subsystem/module*/
    uint8_t log; /**< log setting of subsystem/module*/
    version_t version; /**< software version of module*/
    void (*callback)(int argc, char *argv[]); ///< callback function to process commands
} subsystem_t;

/// system module structure
typedef struct {
    int last_subsys; /**< index of the last subsytem in the array */
    struct {
        uint8_t echo:1;
        uint8_t mute:1;
        uint8_t :6;
    } settings; /**< binary settings for subsystem management*/
    int id; /**< subsystem ID of the subsystem management module*/
    subsystem_t subsystem[MAX_SUBSYSTEMS]; /**< subsystem array */
} system_t;
// array of subsystem information
system_t sys;


#ifndef SUBSYS_MAX_CHARS_PER_CALL
#define SUBSYS_MAX_CHARS_PER_CALL 20
#endif

// log level name to be used in the message header
char * priority_level_name[] = {
    "FORCE",
    "ERR1",
    "ERR2",
    "ERR3",
    "WNG1",
    "WNG2",
    "WNG3",
    "MSG1",
    "MSG2",
    "MSG3",
    "ALL_"
};

// simple functions to access different subsystem info

char *GetSubsystemName(uint8_t subsystem_id) {
    return (sys.subsystem[subsystem_id].name == 0) ? "uninitialized" : sys.subsystem[subsystem_id].name;
}

void SystemInit(void) {
    static uint8_t init_flag = 0;
    if (init_flag) return;
    init_flag = 1;
    sys.last_subsys = -1;
    sys.settings.echo = 0;
    sys.settings.mute = 0;
    // Register the module with the system and give it the name "SYS"
    sys.id = Subsystem_Init("SYS", (version_t)SYSTEM_VERSION, SystemCallback);
    // Register a receiver for the subsystem module
    UART_RegisterReceiver(SUBSYS_UART, ReceiveChar);
}

uint8_t Subsystem_Init(char * name, version_t version, void (*callback)(int argc, char *argv[])) {
    SystemInit();
    sys.last_subsys++;
    if (sys.last_subsys < 0 || sys.last_subsys >= MAX_SUBSYSTEMS) return 0xFF;
    sys.subsystem[sys.last_subsys].log = 1;
    sys.subsystem[sys.last_subsys].name = name;
    sys.subsystem[sys.last_subsys].version.word = version.word;
    sys.subsystem[sys.last_subsys].callback = callback;
    return sys.last_subsys;
}

void Subsystem_RegisterCallback(uint8_t subsystem_id, void (*callback)(int argc, char *argv[])) {
    sys.subsystem[subsystem_id].callback = callback;
}

enum com_state {
    IDLE, WRITING
};

char receive_buffer[RECEIVE_MAX_LENGTH];

void ReceiveChar(char c) {
    //static char receive_buffer[LOG_RECEIVE_MAX_LENGTH];
    static enum com_state receive_state = IDLE;
    static int length;
    if(sys.settings.echo) {
        UART_WriteByte(SUBSYS_UART, c);
        if(c == RECEIVE_STOP_CHAR) UART_WriteByte(SUBSYS_UART, '\n');
    }
    if (receive_state == IDLE) {
        if (c == RECEIVE_START_CHAR) {
            receive_state = WRITING;
            length = 0;
        }
    } else {
        receive_buffer[length] = c;
        if (c == RECEIVE_STOP_CHAR) {
            // process received message
            ProcessMsg();
            receive_state = IDLE;
        } else {
            if (c == ASCII_BACKSPACE) {
                if (length) length--;
                else receive_state = IDLE;
            } else {
                length++;
                if (length >= RECEIVE_MAX_LENGTH) receive_state = IDLE;
            }
        }
    }
}

void ProcessMsg(void) {
    volatile int16_t i = 0;
    char *name;
    char *argv[RECEIVE_MAX_ARGC];
    int argc = 0;
    // get the name
    name = &receive_buffer[0];

    while (receive_buffer[i] != RECEIVE_STOP_CHAR) {
        // if the index is greater than length then break;
        if (i >= RECEIVE_MAX_LENGTH) break;
        if (receive_buffer[i] == ' ') {
            receive_buffer[i] = 0;
            break;
        }
        i++;
    }
    // receive the arguments
    argv[argc] = &receive_buffer[++i];
    while (receive_buffer[i] != RECEIVE_STOP_CHAR) {
        // if the index is greater than length then break;
        if (i >= RECEIVE_MAX_LENGTH) break;
        i++;
        if (receive_buffer[i] == ' '  || receive_buffer[i] == RECEIVE_STOP_CHAR) {
            argc++;
            if(receive_buffer[i] == RECEIVE_STOP_CHAR) {
                receive_buffer[i] = 0;
                break;
            }else {
                receive_buffer[i] = 0;
                argv[argc] = &receive_buffer[++i];
            }
        }
    }

    for (i = 0; i <= sys.last_subsys; i++) {
        if (strcasecmp(sys.subsystem[i].name, name) == 0) {
            if (sys.subsystem[i].callback) {
                sys.subsystem[i].callback(argc, argv);
            }
        }
    }
}

int16_t FlagStr2Int(char * ptr) {
    // make sure a number is after the flag
    if (*ptr >= '0' && *ptr <= '9') {
        // can accept 0 to 99
        if (*(ptr + 1) >= '0' && *(ptr + 1) <= '9') {
            return (*ptr - '0')*10 + *(ptr + 1) - '0';
        } else return *ptr - '0';
    } else {
        // code could be added here to allow flags to use level
        // names instead of numbers
        LogMsg(sys.id, "Invalid level flag format");
        return 0;
    }
}
/*
void LogVersions(void) {
    uint16_t i;
    char * ptr;
    version_mmb_t * v;
    LogMsg(SYSTEM, FORCE_LOG, "Subsystem versions:");
    for (i = 0; i < NUM_SUBSYSTEMS; i++) {
        ptr = subsystem[i].name;
        if (ptr) {
            v = &subsystem[i].version.v;
            LogStr("%s %u.%u.%u\r\n", ptr, v->major, v->minor, v->build);
        }
    }
}

void LogSubsystem(int16_t s) {
    char *ptr, *ptr2;

    ptr = GetSubsystemName((enum sys_index)s); // cast to sys_index not required but avoids warning on strict compilers
    if(subsystem[s].log_priority == FORCE_LOG) ptr2 = "OFF";
    else ptr2 = priority_level_name[subsystem[s].log_priority];
    LogStr("%u %s, level: %s\r\n", s, ptr, ptr2);
}

void LogLevels(void) {
    enum priority_level level;

    LogMsg(SYSTEM, FORCE_LOG, "Log level listing:");
    for (level = FORCE_LOG; level <= EVERYTHING; level++) {
        LogStr("%d %s\r\n", level, priority_level_name[level]);
    }
}

void LogSetAllLevels(enum priority_level level) {
    uint16_t i;
    for (i = 0; i < NUM_SUBSYSTEMS; i++) {
        subsystem[i].log_priority = level;
    }
}

void LogSetGlobalLevel(enum priority_level level) {
    global_log_level = level;
}

void LogSubsystemLevel(enum sys_index sys, enum priority_level level) {
    subsystem[sys].log_priority = level;
}
 */
#define LogChar(c) UART_WriteByte(SUBSYS_UART,c)

void LogHeader(uint8_t subsystem_id);

void LogStr(char * str, ...) {
    // variable argument list type
    va_list vars;
    // initialize the variable argument list pointer by specifying the
    // input argument immediately preceding the variable list
    va_start(vars, str);
    // use Push_vprintf to log to LOG_BUF
    UART_vprintf(SUBSYS_UART, str, vars);
    va_end(vars);
}

void LogMsg(uint8_t subsystem_id, char * str, ...) {
    // variable argument list type
    va_list vars;
    if(subsystem_id > sys.last_subsys) return;
    if (sys.subsystem[subsystem_id].log & sys.settings.mute == 0) {
        // output header
        Log_Header(subsystem_id);
        // initialize the variable argument list pointer by specifying the
        // input argument immediately preceding the variable list
        va_start(vars, str);
        // use Push_vprintf to log to LOG_BUF
        UART_vprintf(SUBSYS_UART, str, vars);
        va_end(vars);
        // new line
        LogChar('\r');
        LogChar('\n');
    }
}

void Log_Header(uint8_t subsystem_id) {
    // format header as "<ms timestamp>_<subsystem name> "
    UART_Printf(SUBSYS_UART, "%l_%s ",TimeNow(), GetSubsystemName(subsystem_id));
}

static void SystemCallback(int argc, char *argv[]) {
    volatile uint8_t i;
    if(argc) {
        if(strcasecmp(argv[0],"list") == 0) {
            for(i = 0; i < MAX_SUBSYSTEMS; i++) {
                if(sys.subsystem[i].name) LogStr("%d: %s\r\n", i, sys.subsystem[i].name);
            }
        }else if(strcasecmp(argv[0],"mute") == 0) {
            LogMsg(sys.id, "muted");
            sys.settings.mute = 1;
        }else if(strcasecmp(argv[0],"unmute") == 0) {
            sys.settings.mute = 0;
            if(argc >= 2 && strcasecmp(argv[1], "all") == 0) {
                for(i = 0; i <= sys.last_subsys; i++) sys.subsystem[i].log = 0;
                LogMsg(sys.id, "unmuted all");
            }else LogMsg(sys.id, "unmuted");
        }else if(strcasecmp(argv[0],"echo") == 0) {
            if(argc >= 2 && strcasecmp(argv[1],"off") == 0) {
                sys.settings.echo = 0;
                LogMsg(sys.id, "echo off");
            }else {
                sys.settings.echo = 1;
                LogMsg(sys.id, "echo on");
            }
        }else{
            //check if the command is actually a subsystem name
            for(i = 0; i <= sys.last_subsys; i++) {
                if(strcasecmp(argv[0], sys.subsystem[i].name) == 0) {
                    if(argc < 2) LogMsg(sys.id, "to few args");
                    else {
                        if(strcasecmp(argv[1],"mute") == 0) {
                            LogMsg(sys.id, "muted");
                            sys.subsystem[i].log = 0;
                        }else if(strcasecmp(argv[1],"unmute") == 0) {
                            sys.subsystem[i].log = 1;
                            LogMsg(sys.id, "unmuted");
                        }else LogMsg(sys.id, "cmd not supported");
                    }
                    return;
                }
            }
            LogMsg(sys.id, "cmd not supported");
        }
    }
}

void Log_EchoOn(void) {
    sys.settings.echo = 1;
}

void Log_EchoOff(void) {
    sys.settings.echo = 0;
}

uint8_t Log_GetEcho(void) {
    if(sys.settings.echo) return 1;
    return 0;
}
